### PR TITLE
[10.x.x] Bugfix 1262887 - Transparent meshes are rendered as opaques when using Custom Render Pass

### PR DESCRIPTION
**Pull request by ellioman**
# Purpose of this PR
Fixes [bug 1262887](https://issuetracker.unity3d.com/issues/urp-transparent-meshes-are-rendered-as-opaques-when-using-lit-shader-with-custom-render-pass).

In January I originally made a PR to fix [bug 1168283](https://issuetracker.unity3d.com/issues/lwrp-alpha-clipping-material-makes-other-materials-look-like-alpha-clipping-when-gameobject-is-shown-in-render-texture). 

The new bug shows that my previous solution didn't work in all cases. An example is if the renderer renders opaque objects with DrawObjectsPass and then transparent objects with a custom renderer feature, the transparent objects would be rendered as opaque as well.

**This PR fixes the issue by:**
- deprecating `_DrawObjectPassData` in our shaders
- adding `_Surface` to our CBuffers, which is either 0 (Opaque) or 1 (Transparent)
- update the OutputAlpha() shader function to take surfaceType as input and use that to output the correct Alpha value
- Update our shaders to send the surfaceTypes to the OutputAlpha() function

# Automated Tests
Added a new test scene **110_URPShadersAlphaOutputCustomFeature**, which is almost exactly the test scene I added when fixing the previous bug except it now uses a custom pass to render the transparent objects.

Results before this PR:
image

Results with this PR:
image

# Yamato**
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fbugfix%252F1262887-transparents-renderered-opaque

### CODE REVIEW

File: TestProjects/UniversalGraphicsTest/Assets/Scenes/110_URPShadersAlphaOutputCustomFeature/RenderTransparents.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+using UnityEngine;
```
Comment (phi-lira) [inline TestProjects/UniversalGraphicsTest/Assets/Scenes/110_URPShadersAlphaOutputCustomFeature/RenderTransparents.cs:1]: I wonder if you couldn't test this by adding the RenderObjects pass feature?

File: com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl [hunk 2]
```diff
@@ -4,21 +4,24 @@
 #include "Packages/com.unity.render-pipelines.universal/ShaderLibrary/SurfaceInput.hlsl"
 
 CBUFFER_START(UnityPerMaterial)
-float4 _BaseMap_ST;
-half4 _BaseColor;
-half _Cutoff;
-half _Glossiness;
-half _Metallic;
+    half _Surface;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl:7]: I'd expect that if you add this at the end you get better padding/layout.

File: com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl [hunk 3]
```diff
@@ -4,21 +4,24 @@
...
+    float4 _BaseMap_ST;
+    half4 _BaseColor;
+    half _Cutoff;
+    half _Glossiness;
+    half _Metallic;
 CBUFFER_END
 
 #ifdef UNITY_DOTS_INSTANCING_ENABLED
 UNITY_DOTS_INSTANCING_START(MaterialPropertyMetadata)
+    UNITY_DOTS_INSTANCED_PROP(float , _Surface)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl:17]: Same here, why not adding at the end?

File: com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl (hunk 2 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl:7]: Another suggestion `_SurfaceType`

File: com.unity.render-pipelines.universal/Shaders/LitInput.hlsl [hunk 4]
```diff
@@ -38,6 +40,7 @@ UNITY_DOTS_INSTANCING_START(MaterialPropertyMetadata)
     //UNITY_DOTS_INSTANCED_PROP(float , _ClearCoatSmoothness) // TODO: enable
 UNITY_DOTS_INSTANCING_END(MaterialPropertyMetadata)
 
+#define _Surface                UNITY_ACCESS_DOTS_INSTANCED_PROP_FROM_MACRO(float4 , Metadata__Surface)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Shaders/LitInput.hlsl:43]:
```suggestion
#define _Surface                UNITY_ACCESS_DOTS_INSTANCED_PROP_FROM_MACRO(float , Metadata__Surface)
```

Comment (ellioman) [reply com.unity.render-pipelines.universal/Shaders/LitInput.hlsl]: Fixed

File: com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl (hunk 3 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl]: Fixed

File: com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl (hunk 2 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/Shaders/BakedLitInput.hlsl]: That would require a material upgrader or having two different parameters for the same thing in our shaders.

File: TestProjects/UniversalGraphicsTest/Assets/Scenes/110_URPShadersAlphaOutputCustomFeature/RenderTransparents.cs (hunk 1 above)
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/Scenes/110_URPShadersAlphaOutputCustomFeature/RenderTransparents.cs]: Was able to. Killed that custom feature.

**Review by fra-frassineti-unity — APPROVED** [review]
Version: 2020.2.0b1.2959
Revision: trunk 4ae7c5e87f33
Built: Tue, 25 Aug 2020 03:44:09 GMT

Test Doc: https://docs.google.com/document/d/14NkqK0MymB0gfKdafHcTIedQoDFFfDK3gLGUK22sIEs/edit?usp=sharing

**Pull request merged**